README.md
# Cube Transition

[![pub package](https://img.shields.io/pub/v/cube_transition.svg)](https://pub.dev/packages/cube_transition)

A 3D Cube transition for your PageView and PageRoute.

<p>
<a href="https://aeyrium.com/" target="_blank"><img src="https://drive.google.com/uc?id=1TP33E66rwZ6wsbaO3vw7PqrMFY2sgZxE" width="500px"
alt="Aeyrium Inc. is a provider of innovative software solutions for the business and commercial aviation"/></a>
</p>
</p>

## Features

- Use the cube transition between the navigation of your pages.
- Add cube transitions to your PageView.
- Add custom transformation in your items of your PageView.


## Getting started

You should ensure that you add the dependency in your flutter project.

```yaml
dependencies:
  cube_transition: "^1.0.0"
```

You should then run `flutter packages upgrade` or update your packages in IntelliJ.

Import the package:

```dart
import 'package:cube_transition/cube_transition.dart';
```


## Example Project

There is a example project in the `example` folder. Check it out. Otherwise, keep reading to get up and running.

## Usage

**Navigator Cube Transition**

```dart
class Sample1 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Navigator Cube Transition'),
      ),
      body: Center(
        child: RaisedButton(
          child: Text('Next Page'),
          onPressed: () {
            Navigator.of(context).push(
              CubePageRoute(
                enterPage: Sample1NextPage(),
                exitPage: this,
                duration: const Duration(milliseconds: 900),
              ),
            );
          },
        ),
      ),
    );
  }
}
```

<p align="center">
  <img src="https://media.giphy.com/media/QaZ24HB4zgl40GMd0k/giphy.gif">
</p>

**PageView Cube Transition (default)**

```dart
class Sample2 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final height = MediaQuery.of(context).size.height / 2;
    return Scaffold(
      appBar: AppBar(
        title: Text('PageView Cube Transition'),
      ),
      body: Center(
        child: SizedBox(
          height: height,
          child: CubePageView(
            children: places
                .map(
                  (item) =>  Image.network(
                        item.url,
                        height: height,
                        fit: BoxFit.cover,
                      ),
                )
                .toList(),
          ),
        ),
      ),
    );
  }
}

```

<p align="center">
  <img src="https://media.giphy.com/media/lnOUMeIqrtosHMc0vM/giphy.gif">
</p>

**PageView Cube Transition Custom** (If you want more control over the items inside the CubePageView)

```dart
class Sample3 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final height = MediaQuery.of(context).size.height / 2;
    return Scaffold(
      appBar: AppBar(
        title: Text('Custom Cube Transition'),
      ),
      body: Center(
        child: SizedBox(
          height: height,
          child: CubePageView.builder(
            itemCount: places.length,
            itemBuilder: (context, index, notifier) {
              final item = places[index];
              final transform = Matrix4.identity();
              final t = (index - notifier).abs();
              final scale = lerpDouble(1.5, 0, t);
              transform.scale(scale, scale);
              return CubeWidget(
                index: index,
                pageNotifier: notifier,
                child: Stack(
                  children: [
                    Image.network(
                      item.url,
                      height: height,
                      fit: BoxFit.cover,
                    ),
                    Transform(
                      alignment: Alignment.center,
                      transform: transform,
                      child: Center(
                        child: Padding(
                          padding: const EdgeInsets.all(12.0),
                          child: Container(
                            decoration: BoxDecoration(boxShadow: [
                              BoxShadow(
                                color: Colors.black45,
                                spreadRadius: 5,
                                blurRadius: 5,
                              ),
                            ]),
                            child: Text(
                              item.name.split('-').join('\n'),
                              textAlign: TextAlign.center,
                              style: Theme.of(context).textTheme.title,
                            ),
                          ),
                        ),
                      ),
                    ),
                  ],
                ),
              );
            },
          ),
        ),
      ),
    );
  }
}

```

<p align="center">
  <img src="https://media.giphy.com/media/SUoqVxRxNLRAaqo6u5/giphy.gif">
</p>



Copyright [Aeyrium Inc](https://aeyrium.com/)

You can follow me on twitter [@diegoveloper](https://www.twitter.com/diegoveloper)


CHANGELOG.md
## [1.0.0 - 1.0.1 - 1.0.2] 

* Initial release.


example/lib/data.dart
class Place {
  final String name;
  final String url;

  const Place({this.name, this.url});
}

const places = [
  const Place(
      name: 'Taj Mahal - India',
      url:
          'https://www.worldatlas.com/r/w728/upload/bb/a2/fa/shutterstock-180918317.jpg'),
  const Place(
      name: 'Colosseum - Italy',
      url:
          'https://www.worldatlas.com/r/w728/upload/55/2c/54/shutterstock-433413835.jpg'),
  const Place(
      name: 'Chichen Itza - Mexico',
      url:
          'https://www.worldatlas.com/r/w728/upload/18/ec/64/shutterstock-356871482.jpg'),
  const Place(
      name: 'Machu Picchu - Peru',
      url:
          'https://www.worldatlas.com/r/w728/upload/ba/e8/c1/shutterstock-168497345.jpg'),
  const Place(
      name: 'Christ the Redeemer - Brazil',
      url:
          'https://www.worldatlas.com/r/w728/upload/a2/b7/90/shutterstock-1283692720-1.jpg'),
  const Place(
      name: 'Petra - Jordan',
      url:
          'https://www.worldatlas.com/r/w728/upload/ba/36/57/shutterstock-1030695895.jpg'),
  const Place(
      name: 'Great Wall of China - China',
      url:
          'https://www.worldatlas.com/r/w728/upload/ce/fc/93/shutterstock-275490581.jpg'),
];


example/lib/main.dart
import 'package:flutter/material.dart';
import 'sample1.dart';
import 'sample2.dart';
import 'sample3.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData.dark(),
      debugShowCheckedModeBanner: false,
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  void _onTap(BuildContext context, Widget page) {
    Navigator.of(context).push(MaterialPageRoute(
      builder: (_) => page,
    ));
  }

  @override
  Widget build(BuildContext context) {
    const separator = const SizedBox(
      height: 15,
    );
    return Scaffold(
      appBar: AppBar(
        title: Text("Examples"),
      ),
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            RaisedButton(
              child: Text("Navigator Cube Transition"),
              onPressed: () => _onTap(context, Sample1()),
            ),
            separator,
            RaisedButton(
              child: Text("PageView Cube Transition"),
              onPressed: () => _onTap(context, Sample2()),
            ),
            separator,
            RaisedButton(
              child: Text("PageView Cube Transition Custom"),
              onPressed: () => _onTap(context, Sample3()),
            ),
          ],
        ),
      ),
    );
  }
}


example/lib/sample1.dart
import 'package:cube_transition/cube_transition.dart';
import 'package:flutter/material.dart';

class Sample1 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Navigator Cube Transition'),
      ),
      body: Center(
        child: RaisedButton(
          child: Text('Next Page'),
          onPressed: () {
            Navigator.of(context).push(
              CubePageRoute(
                enterPage: Sample1NextPage(),
                exitPage: this,
                duration: const Duration(milliseconds: 900),
              ),
            );
          },
        ),
      ),
    );
  }
}

class Sample1NextPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.brown,
      appBar: AppBar(
        title: Text('Page 2'),
      ),
      body: Center(
        child: FlutterLogo(
          size: MediaQuery.of(context).size.width / 2,
        ),
      ),
    );
  }
}


example/lib/sample2.dart
import 'package:cube_transition/cube_transition.dart';
import 'package:flutter/material.dart';
import 'data.dart';

class Sample2 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final height = MediaQuery.of(context).size.height / 2;
    return Scaffold(
      appBar: AppBar(
        title: Text('PageView Cube Transition'),
      ),
      body: Center(
        child: SizedBox(
          height: height,
          child: CubePageView(
            children: places
                .map(
                  (item) => Stack(
                    children: [
                      Image.network(
                        item.url,
                        height: height,
                        fit: BoxFit.cover,
                      ),
                      Positioned(
                        left: 0,
                        bottom: 0,
                        child: Padding(
                          padding: const EdgeInsets.all(12.0),
                          child: Container(
                            decoration: BoxDecoration(boxShadow: [
                              BoxShadow(
                                color: Colors.black45,
                                spreadRadius: 5,
                                blurRadius: 5,
                              ),
                            ]),
                            child: Text(
                              item.name,
                              style: Theme.of(context).textTheme.title,
                            ),
                          ),
                        ),
                      ),
                    ],
                  ),
                )
                .toList(),
          ),
        ),
      ),
    );
  }
}


lib/src/cube_page_view.dart
import 'dart:math';
import 'dart:ui';
import 'package:flutter/material.dart';

/// Signature for a function that creates a widget for a given index in a [CubePageView]
///
/// Used by [CubePageView.builder] and other APIs that use lazily-generated widgets.
///
typedef CubeWidgetBuilder = CubeWidget Function(
    BuildContext context, int index, double pageNotifier);

/// This Widget has the [PageView] widget inside.
/// It works in two modes :
///   1 - Using the default constructor [CubePageView] passing the items in `children` property.
///   2 - Using the factory constructor [CubePageView.builder] passing a `itemBuilder` and `itemCount` properties.

class CubePageView extends StatefulWidget {
  /// Called whenever the page in the center of the viewport changes.
  final ValueChanged<int> onPageChanged;

  /// An object that can be used to control the position to which this page
  /// view is scrolled.
  final PageController controller;

  /// Builder to customize your items
  final CubeWidgetBuilder itemBuilder;

  /// The number of items you have, this is only required if you use [CubePageView.builder]
  final int itemCount;

  /// Widgets you want to use inside the [CubePageView], this is only required if you use [CubePageView] constructor
  final List<Widget> children;

  /// Creates a scrollable list that works page by page from an explicit [List]
  /// of widgets.
  const CubePageView({
    Key key,
    this.onPageChanged,
    this.controller,
    @required this.children,
  })  : itemBuilder = null,
        itemCount = null,
        assert(children != null),
        super(key: key);

  /// Creates a scrollable list that works page by page using widgets that are
  /// created on demand.
  ///
  /// This constructor is appropriate if you want to customize the behavior
  ///
  /// Providing a non-null [itemCount] lets the [CubePageView] compute the maximum
  /// scroll extent.
  ///
  /// [itemBuilder] will be called only with indices greater than or equal to
  /// zero and less than [itemCount].

  CubePageView.builder({
    Key key,
    @required this.itemCount,
    @required this.itemBuilder,
    this.onPageChanged,
    this.controller,
  })  : this.children = null,
        assert(itemCount != null),
        assert(itemBuilder != null),
        super(key: key);

  @override
  _CubePageViewState createState() => _CubePageViewState();
}

class _CubePageViewState extends State<CubePageView> {
  final _pageNotifier = ValueNotifier(0.0);
  PageController _pageController;

  void _listener() {
    _pageNotifier.value = _pageController.page;
  }

  @override
  void initState() {
    _pageController = widget.controller ?? PageController();
    WidgetsBinding.instance.addPostFrameCallback((_) {
      _pageController.addListener(_listener);
    });
    super.initState();
  }

  @override
  void dispose() {
    _pageController.removeListener(_listener);
    _pageController.dispose();
    _pageNotifier.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Material(
      color: Colors.transparent,
      child: Center(
        child: ValueListenableBuilder<double>(
          valueListenable: _pageNotifier,
          builder: (_, value, child) => PageView.builder(
            controller: _pageController,
            onPageChanged: widget.onPageChanged,
            physics: const ClampingScrollPhysics(),
            itemCount: widget.itemCount ?? widget.children.length,
            itemBuilder: (_, index) {
              if (widget.itemBuilder != null)
                return widget.itemBuilder(context, index, value);
              return CubeWidget(
                child: widget.children[index],
                index: index,
                pageNotifier: value,
              );
            },
          ),
        ),
      ),
    );
  }
}

/// This widget has the logic to do the 3D cube transformation
/// It only should be used if you use [CubePageView.builder]

class CubeWidget extends StatelessWidget {
  /// Index of the current item
  final int index;

  /// Page Notifier value, it comes from the [CubeWidgetBuilder]
  final double pageNotifier;

  /// Child you want to use inside the Cube
  final Widget child;

  const CubeWidget({
    Key key,
    @required this.index,
    @required this.pageNotifier,
    @required this.child,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final isLeaving = (index - pageNotifier) <= 0;
    final t = (index - pageNotifier);
    final rotationY = lerpDouble(0, 90, t);
    final opacity = lerpDouble(0, 1, t.abs()).clamp(0.0, 1.0);
    final transform = Matrix4.identity();
    transform.setEntry(3, 2, 0.003);
    transform.rotateY(-degToRad(rotationY));
    return Transform(
      alignment: isLeaving ? Alignment.centerRight : Alignment.centerLeft,
      transform: transform,
      child: Stack(
        children: [
          child,
          Positioned.fill(
            child: Opacity(
              opacity: opacity,
              child: Container(
                child: Container(
                  color: Colors.black87,
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

num degToRad(num deg) => deg * (pi / 180.0);
num radToDeg(num rad) => rad * (180.0 / pi);


lib/src/cube_page_route.dart
import 'package:flutter/material.dart';
import 'dart:math';

/// This is a Custom page route we use to simulate the 3D Cubic transition from page to page
/// It should be used with the [Navigator]
///
/// This was taken from: https://stackoverflow.com/questions/59649977/flutter-transition-between-pages-with-effect-3d
///
class CubePageRoute extends PageRouteBuilder {
  /// The duration of the transition
  final Duration duration;

  /// The next page we will show
  final Widget enterPage;

  /// The current page we will dismiss
  final Widget exitPage;

  /// Background color used when the transition is running , it's [Colors.black] as default
  final Color backgroundColor;

  CubePageRoute({
    this.exitPage,
    this.enterPage,
    this.backgroundColor = Colors.black,
    this.duration = const Duration(
      milliseconds: 500,
    ),
  }) : super(
          transitionDuration: duration,
          pageBuilder: (context, animation, secondaryAnimation) => enterPage,
          transitionsBuilder: (context, animation, secondaryAnimation, child) {
            return Container(
              color: backgroundColor,
              child: Stack(
                children: <Widget>[
                  SlideTransition(
                    position: Tween<Offset>(
                      begin: Offset.zero,
                      end: Offset(-1.0, 0.0),
                    ).animate(animation),
                    child: Transform(
                      transform: Matrix4.identity()
                        ..setEntry(3, 2, 0.003)
                        ..rotateY(pi / 2 * animation.value),
                      alignment: FractionalOffset.centerRight,
                      child: exitPage,
                    ),
                  ),
                  SlideTransition(
                    position: Tween<Offset>(
                      begin: Offset(1.0, 0.0),
                      end: Offset.zero,
                    ).animate(animation),
                    child: Transform(
                      transform: Matrix4.identity()
                        ..setEntry(3, 2, 0.003)
                        ..rotateY(pi / 2 * (animation.value - 1)),
                      alignment: FractionalOffset.centerLeft,
                      child: enterPage,
                    ),
                  )
                ],
              ),
            );
          },
        );
}


example/lib/sample3.dart
import 'dart:ui';

import 'package:cube_transition/cube_transition.dart';
import 'package:flutter/material.dart';
import 'data.dart';

class Sample3 extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final height = MediaQuery.of(context).size.height / 2;
    return Scaffold(
      appBar: AppBar(
        title: Text('Custom Cube Transition'),
      ),
      body: Center(
        child: SizedBox(
          height: height,
          child: CubePageView.builder(
            itemCount: places.length,
            itemBuilder: (context, index, notifier) {
              final item = places[index];
              final transform = Matrix4.identity();
              final t = (index - notifier).abs();
              final scale = lerpDouble(1.5, 0, t);
              transform.scale(scale, scale);
              return CubeWidget(
                index: index,
                pageNotifier: notifier,
                child: Stack(
                  children: [
                    Image.network(
                      item.url,
                      height: height,
                      fit: BoxFit.cover,
                    ),
                    Transform(
                      alignment: Alignment.center,
                      transform: transform,
                      child: Center(
                        child: Padding(
                          padding: const EdgeInsets.all(12.0),
                          child: Container(
                            decoration: BoxDecoration(boxShadow: [
                              BoxShadow(
                                color: Colors.black45,
                                spreadRadius: 5,
                                blurRadius: 5,
                              ),
                            ]),
                            child: Text(
                              item.name.split('-').join('\n'),
                              textAlign: TextAlign.center,
                              style: Theme.of(context).textTheme.title,
                            ),
                          ),
                        ),
                      ),
                    ),
                  ],
                ),
              );
            },
          ),
        ),
      ),
    );
  }
}
